### PR TITLE
Try using FindOpenMP instead of figuring out flags manually

### PR DESCRIPTION
**Pull request by masterleinad**
Trying to address #3256 by relying on `CMake`'s `FindOpenMP`.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
This pull currently doesn't add any linker flags but passes our regular CI. It would be good to check this on a somewhat wider ranger of configurations.

**Comment by ndellingwood** [comment]
I configured and built this PR with intel/17.0.4 and intel/19.0.3 each with cmake/3.16.0 and cmake/3.19.3, there were no issues in any of these cases

**Comment by masterleinad** [comment]
I think the interesting cases to test with are `CLANG_CRAY` and `CLANG_MSVC`.

**Comment by ndellingwood** [comment]
I'm not able to test clang_msvc, here is from my attempt to test with clang_cray that failed to configure (I don't have much experience with that env, so I could be doing something wrong):

Load cray computing env
```
module load cce/11.0.2 cmake/3.16.2

clang++ --version
Cray clang version 11.0.2  (7cc96bf2bf88b71c173230d893018a3f38fbbb09)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

Configure
Kokkos SHA 4a29d0a124bc8b0d7bc818a2069b50d7be0a748b
```
cmake -DCMAKE_CXX_COMPILER=clang++ -DKokkos_ARCH_HSW=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_TESTS=ON $KOKKOS_PATH
```

Configure output:
```
ndellin@voltrino:~/kokkos/testing/Voltrino-HSW-CCE> cmake -DCMAKE_CXX_COMPILER=clang++ -DKokkos_ARCH_HSW=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_TESTS=ON $KOKKOS_PATH 2>&1 | tee config.out
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is Clang 11.0.0
-- Check for working CXX compiler: /opt/cray/pe/cce/11.0.2/cce-clang/x86_64/bin/clang++
-- Check for working CXX compiler: /opt/cray/pe/cce/11.0.2/cce-clang/x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
CMake Warning at cmake/kokkos_check_env.cmake:10 (MESSAGE):
  CRAYPE_LINK_TYPE is not set.  Linking is likely to fail unless this is set
  to 'dynamic'
Call Stack (most recent call first):
  CMakeLists.txt:162 (INCLUDE)


CMake Error at /projects/Voltrino/hpcsoft/cle7.0/common/cmake/3.16.2/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /projects/Voltrino/hpcsoft/cle7.0/common/cmake/3.16.2/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /projects/Voltrino/hpcsoft/cle7.0/common/cmake/3.16.2/share/cmake-3.16/Modules/FindOpenMP.cmake:511 (find_package_handle_standard_args)
  cmake/kokkos_enable_devices.cmake:47 (FIND_PACKAGE)
  cmake/kokkos_tribits.cmake:229 (INCLUDE)
  CMakeLists.txt:170 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```

Is there anything obvious that I may be doing wrong on my end?

**Comment by masterleinad** [comment]
I almost forgot that I can check myself for clang_cray. On `spock`, I get
```
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is Clang 11.0.0
-- Check for working CXX compiler: /opt/cray/pe/craype/2.7.6/bin/CC
-- Check for working CXX compiler: /opt/cray/pe/craype/2.7.6/bin/CC - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
CMake Warning at cmake/kokkos_check_env.cmake:10 (MESSAGE):
  CRAYPE_LINK_TYPE is not set.  Linking is likely to fail unless this is set
  to 'dynamic'
Call Stack (most recent call first):
  CMakeLists.txt:162 (INCLUDE)


-- Found OpenMP_CXX: -fopenmp (found version "5.0") 
-- Found OpenMP: TRUE (found version "5.0")  
-- Using -std=gnu++14 for C++14 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: Kokkos::OpenMP
--       Host Serial: NONE
-- 
-- Architectures:
-- Found TPLLIBDL: /usr/lib64/libdl.so  
-- Kokkos Devices: OPENMP, Kokkos Backends: OPENMP
-- Configuring done
-- Generating done
-- Build files have been written to: /ccs/home/darndt/kokkos/build_cray
```
so that seems to work. 
@ndellingwood Can you use `OpenMP` successfully on your system with `develop` instead of this pull request?

**Comment by ndellingwood** [comment]
@masterleinad I'm able to get the develop branch to configure,

```
ndellin@voltrino:~/kokkos/testing/Voltrino-HSW-CCE> cmake -DCMAKE_CXX_COMPILER=clang++ -DKokkos_ARCH_HSW=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_TESTS=ON $KOKKOS_PATH
-- Setting default Kokkos CXX standard to 14
-- The CXX compiler identification is Clang 11.0.0
-- Check for working CXX compiler: /opt/cray/pe/cce/11.0.2/cce-clang/x86_64/bin/clang++
-- Check for working CXX compiler: /opt/cray/pe/cce/11.0.2/cce-clang/x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
CMake Warning at cmake/kokkos_check_env.cmake:10 (MESSAGE):
  CRAYPE_LINK_TYPE is not set.  Linking is likely to fail unless this is set
  to 'dynamic'
Call Stack (most recent call first):
  CMakeLists.txt:162 (INCLUDE)


-- Using -std=gnu++14 for C++14 extensions as feature
-- Built-in Execution Spaces:
--     Device Parallel: NoTypeDefined
--     Host Parallel: Kokkos::OpenMP
--       Host Serial: NONE
-- 
-- Architectures:
--  HSW
-- Found TPLLIBDL: /usr/lib64/libdl.so  
-- Sources TestOpenMP.cpp
-- Kokkos Devices: OPENMP, Kokkos Backends: OPENMP
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ndellin/kokkos/testing/Voltrino-HSW-CCE
```

However I am running into build errors in this case:
```
ld.lld: error: unable to find library -lcraymp
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by crtrott** [comment]
DId you test CUDA + OpenMP?

**Comment by crtrott** [comment]
I did some testing with ICPX for simple OpenMP run, OpenMPTarget, and CUDA+OpenMP even with nvc++ and it all works. So this appears to be good on that front. Still we need to resolve the cray stuff.

**Comment by crtrott** [comment]
@masterleinad is that still WIP? Or does this work now?

**Comment by masterleinad** [comment]
> @masterleinad is that still WIP? Or does this work now?

I didn't change anything but I would think that the error in https://github.com/kokkos/kokkos/pull/4105#issuecomment-881727991 is more an issue with the machine being misconfigured. I would be interesting to hear if the problem persists.

**Comment by masterleinad** [comment]
Just as another `cray` data point: This works fine on `spock` as well.

**Review by cz4rs — APPROVED** [review]
Looks good to me and rebases cleanly on top of `develop` 👍

**Comment by crtrott** [comment]
@ndellingwood which machine was this failing on and is it still an issue?

**Comment by ndellingwood** [comment]
@crtrott after pulling in an update of this PR I'm still seeing the same issues on Voltrino as here: https://github.com/kokkos/kokkos/pull/4105#issuecomment-881719522

I'm not very familiar with this toolchain, so it could be something wrong I'm doing?

File: cmake/kokkos_enable_devices.cmake [hunk 1]
```diff
@@ -44,43 +44,11 @@ ELSE()
...
-      XL         -qsmp=omp
-      DEFAULT    -fopenmp
-    )
-    COMPILER_SPECIFIC_LIBS(
-      AppleClang -lomp
-    )
-  ENDIF()
+  FIND_PACKAGE(OpenMP REQUIRED)
+  STRING(REPLACE ";" " " OpenMP_CXX_FLAGS "${OpenMP_CXX_FLAGS}")
+  COMPILER_SPECIFIC_FLAGS(
```
Comment (junghans) [inline cmake/kokkos_enable_devices.cmake:49]: I don't think compiler flags aren't enough as some implementations need `-lomp`, I guess something like:
```
KOKKOS_LINK_TPL(kokkoscore PUBLIC IMPORTED_NAME OpenMP::OpenMP_CXX OPENMP)
```
would do it.

**Comment by masterleinad** [comment]
I updated this pull request to just use
```
IF (Kokkos_ENABLE_OPENMP)
  find_package(OpenMP REQUIRED)
  KOKKOS_EXPORT_CMAKE_TPL(OpenMP REQUIRED)
  target_link_libraries(kokkoscore PUBLIC OpenMP::OpenMP_CXX)
ENDIF()
```
which is the same approach we use for `desul`. This would also force `Trilinos` to use the same path instead of
```
KOKKOS_LINK_TPL(kokkoscore PUBLIC IMPORTED_NAME OpenMP::OpenMP_CXX OPENMP)
```
which doesn't do anything for `Trilinos`.

**Review by junghans — APPROVED** [review]

**Comment by junghans** [comment]
I guess that would also help in #5184

**Comment by masterleinad** [comment]
One disadvantage of this approach would be that using different compilers for Kokkos and downstream applications causes issues when using `OpenMP` since the compiler flags for the found target might be different. On the other hand, I wouldn't expect that to be very stable anyway.

**Review by nmm0 — APPROVED** [review]
Looks good to me

**Comment by aminiussi** [comment]
Hi, correct me if I'm wrong, but this fix is not in 3.7.1 right ?
Thanks

**Comment by masterleinad** [comment]
> Hi, correct me if I'm wrong, but this fix is not in 3.7.1 right ?

Yes, this feature is only merged to the 4.0.00 release branch.

**Comment by aminiussi** [comment]
Thanks. Is there a roadmap page somewhere ?

**Comment by masterleinad** [comment]
> Thanks. Is there a roadmap page somewhere ?

We normally only do release notes (https://github.com/kokkos/kokkos/issues/5439, https://github.com/kokkos/kokkos/blob/master/CHANGELOG.md).

**Pull request merged**